libyul/optimiser/Rematerialiser.cpp

```diff
@@ -86,13 +86,17 @@ void Rematerialiser::visit(Expression& _e)
 			)
 			{
 				assertThrow(m_referenceCounts[name] > 0, OptimizerException, "");
-				for (auto const& ref: m_references.forward[name])
-					assertThrow(inScope(ref), OptimizerException, "");
-				// update reference counts
-				m_referenceCounts[name]--;
-				for (auto const& ref: ReferencesCounter::countReferences(*value.value))
-					m_referenceCounts[ref.first] += ref.second;
-				_e = (ASTCopier{}).translate(*value.value);
+				bool allInScope = true;
```

Comment:
```suggestion
				bool allInScope = fold(m_references[name], true, [this](auto&& _acc, auto const& _ref) { return _acc && inScope(_ref); });
				solAssert(allInScope, "");
```

Reply:
Loop can be removed then, but fine with either if you want to keep it.

Reply:
``fold`` is a bit harsh... I could add ``std::ranges::all_of`` to the cxx20 header, though...

Reply:
cpp20 also has a better version of `accumulate`